query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/converters/PinotFunctionConverter.java

```diff
@@ -52,6 +57,16 @@ public String convert(
     }
   }
 
+  private String functionToStringForAvgRate(
+      Function function, java.util.function.Function<Expression, String> argumentConverter) {
+    String argumentString =
+        function.getArgumentsList().stream()
+            .map(argumentConverter::apply)
+            .collect(Collectors.joining(","));
```

Comment:
So the form expected by QS is something like AVG_RATE(colName, divisor)? That doesn't seem quite right still. Who is responsible for calculating that divisor (which would be `aggregationIntervalDuration / rateIntervalDuration`)? If we're really abstracting it, then I would expect QS to be doing that conversion.

Reply:
Yes. That would be a follow up task to clean the implementation in other repos.

Reply:
@sarthak77 I also think, we should do the full conversation in QS. Are we not getting the required information in the QS request to calculate divisor?

Reply:
@aaron-steinfeld @sarthak77 I relooked this again, and here are my thoughts.
As you mentioned, for divisor, it requires two values - `aggregationIntervalDuration and rateIntervalDuration`. 
There are two types of queries we have.
- Non-time series (aggregation query over entire time range filter)
For this, aggregationIntervalDuration interval will be the entire duration window which is (start_time, end_time). If we do at the QS level this needs to be extracted from time filters which can be using any attributes. (BACKEND.starttime, BACKEND.endTime, SERVICE.starttime / SERVICE.endtime or API.starttime / API.endtime)
- Time series based aggregation
Fr this, aggregationIntervalDuration interval needs to be derived from dateTimeConvert function from groupBy clause at QS. Gateway's `time_aggregation` is converted to `dateTimeConvert` of QS.

The gateway knows both time aggregation and (starttime, endtime) window of query, and handles the query transformation. I think it should. So, QS should expose avg_rate with required args like any other aggregation function. And, we can have either of 
- avg_rate(col, divisor) or
- avg_rate(col, aggrInterval, rateInterval)

Reply:
We do have all the information, we just need to extract it from the query. It's a bit more challenging to do that at the QS level, but not impossible by any means - as described by @kotharironak.

Requiring either `divisor` or `aggrInterval, rateInterval` is just putting more responsibility on the caller because of a weak interface against QS. All of the problems come back to our current interface. A wholesale change there isn't necessarily feasible and certainly not within scope, but let's describe and then we can figure out a reasonable solution.

For aggregates:
- We need the time range, which IMO should have always been a top level argument in QS. There's no guarantee that other non-pinot implementations treat time range as just another column like we're assuming here. For example, what would this look like if the query were going to prometheus?
For time series:
- Basically the same argument, but extending it to the series interval. datetimeconvert never should have been specified on the gateway service side, the time series interval should have been a top level arg.


So what would I want to do with enough time (and I think we'll need to do this to support a metric store, anyway): 
New QS APIs! I think I'd argue for two top level APIs, one for aggregates and a separate one for time series.

My suggestion for now:
If we agree that new APIs will be needed in the fairly near term, then I'd go with whatever builds towards that. In my mind, we'd eventually want only the single rate interval argument, so if we have to do a bit of hackishness to extract the extra info on the QS side for now, so be it - that would go away once we get the new interfaces. If we think the interface changes will be even larger and that the code's going to be replaced no matter how we specify the arguments, then we can go with whatever's quickest.

Reply:
Yes, I agree that we should have `start_time`, `end_time`, and `interval`(if time-series query) information at the top-level argument in a QS request. It needs in all metric store queries (else it has to derive internally from filter clause and group by clause accordingly for aggregate and time-series request).

It also makes sense to go with separate APIs for aggregate and time series. However, I think some of the details will be clear once we start working on QS to MetricStore transformation. So, we can define the interfaces accordingly (should be soon, once we are done with the ingestion pipeline).

So, I think, for now, as you suggested, we can handle this at the QS layer by extracting information (start_time, end_time, and interval) from the original request, and storing it in the execution context. Once, this is available as part of interfaces, this code will go away.

> what would this look like if the query were going to prometheus?

One of the example query. As shown below the range (start_time, end_time) will be converted to range selector operator `[]` - `[3600000ms]`
```
http://localhost:9090/api/v1/query?query=sum by (backend_id) 
(delta(num_calls{tenant_id="__default", backend_id='27857c9c-f6fa-30fe-ac1d-48fd07374775'}[3600000ms]))&time=1631178754903
```

Reply:
Right - so we'll need this information extracted from the query to go to prometheus anyway - we can think of this as a little backwards compatibility hack.